Kafka client direct consumer

mmaslankaprv opened the pull request:
Simple direct Kafka consumer. It allows user to subscribe to a list of topic partitions. The consumer maintains broker fetch session, it reacts for the leadership changes and metadata updates. Fetch responses from all the brokers are exposed through `fetch_next()`. The fetch_next() returns whenever any ready to be consumer are available.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

dotnwat commented on src/v/kafka/client/sequential_consumer.h:
```diff
@@ -0,0 +1,140 @@
+};
+
+/**
+ * Simple sequential Kafka consumer. It allows user to subscribe to a list of
```
where does the term `sequential` come from?

rockwotj replied:
I usually see the terms direct consuming vs group consuming.

This a direct consumer, so maybe that's a better term? Franz uses it IIRC

dotnwat commented on src/v/kafka/client/sequential_consumer.h:
```diff
@@ -0,0 +1,140 @@
+ *
+ * Right now the consumer uses very simple buffering strategy where only one
+ * FetchResponse is buffered per broker. After the response is consumed through
+ * `fetch_next()` it triggers a fetch request.
```
👍 it might be useful to know what strategies other clients use. maybe they also use this simple buffering strategy too? would it be more flexible if the user of the consumer provided the buffers or would the strategy still be same?

dotnwat reviewed (commented):
interface looks pretty good to me.

mmaslankaprv committed:
kc/cluster: expose logger from client cluster

It is convenient to use client_id prefixed logger in higher level client
classes.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
kc/direct_consumer: introduced direct consumer class

Introduced a class that allow callers to fetch data from multiple
topic/partitions that are manually assigned to the consumer. The
consumer handles the fetch requests and offset resets.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68670-->
<!--bfe1671d-7613-4fe4-86bf-ab543148f61b-->
<!--title-->
## Retry command for [Build#68670](https://buildkite.com/redpanda/redpanda/builds/68670)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_relaxed_acks@{"write_caching":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e829921a-dcda-4bf1-aaab-38c12908a3b6-->
<!--TEST RESULTS-->
<!--0c76e92b-e2fb-4010-abc3-470a8d27d615-->
<!--TEST RESULTS-->
<!--bd23cc5b-3466-4cfc-95d4-58ca06229e76-->
<!--TEST RESULTS-->
<!--6f775194-96b3-41b0-987e-b36de81bdd7d-->
<!--TEST RESULTS-->
<!--d660d22a-97b6-4fff-b940-60dc9c2d2e38-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68670">build#68670</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68670#0197f0c0-3f14-4ae4-bf82-9d0fd1bfed57 | FLAKY         | 19/21    | upstream reliability is '98.55491329479769'. current run reliability is '90.47619047619048'. drift is 8.07872 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68670#0197f0c0-3f16-4f0d-8aa7-745d96d3f779 | FLAKY         | 20/21    | upstream reliability is '99.32659932659934'. current run reliability is '95.23809523809523'. drift is 4.0885 and the allowed drift is set to 50. The test should PASS  |
| SimpleEndToEndTest       | test_relaxed_acks    | {"write_caching": false}                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68670#0197f0c0-3f12-4597-b114-ce6823536c46 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68914">build#68914</a></summary>


| test_class                                   | test_method                                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:---------------------------------------------|:---------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| cluster_metadata_uploader_fixture            | test_upload_in_term                          |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/68914#019808ed-f68c-4cc2-9360-ad21a03f2687 | FAIL          | 0/1      |          |
| src/v/kafka/client/test/direct_consumer_test | src/v/kafka/client/test/direct_consumer_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/68914#019808ed-f68c-4cc2-9360-ad21a03f2687 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68924">build#68924</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68924#019809f7-3adf-4163-8116-103be73a6c2b | FLAKY         | 18/21    | upstream reliability is '97.45762711864407'. current run reliability is '85.71428571428571'. drift is 11.74334 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69009">build#69009</a></summary>


| test_class                                     | test_method                                    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:-----------------------------------------------|:-----------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| src/v/kafka/client/test/cluster_test_with_mock | src/v/kafka/client/test/cluster_test_with_mock |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/69009#01980ee8-bfe0-46f4-88d8-d0cd578ee417 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69092">build#69092</a></summary>


| test_class                   | test_method                   | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:------------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest    | test_read_chunks              |                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a11b-4505-85ab-ce17508299c0 | FLAKY         | 20/21    | upstream reliability is '96.9047619047619'. current run reliability is '95.23809523809523'. drift is 1.66667 and the allowed drift is set to 50. The test should PASS   |
| CloudStorageTimingStressTest | test_cloud_storage            | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#0198137d-1db6-4bab-bb9e-99ab52542e99 | FLAKY         | 19/21    | upstream reliability is '99.16434540389972'. current run reliability is '90.47619047619048'. drift is 8.68815 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "gzip"}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#0198137d-1db7-42e0-a4a8-04a257d4e808 | FLAKY         | 20/21    | upstream reliability is '94.23076923076923'. current run reliability is '95.23809523809523'. drift is -1.00733 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "gzip"}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a118-47bf-b6e1-9110ec117d74 | FLAKY         | 20/21    | upstream reliability is '94.23076923076923'. current run reliability is '95.23809523809523'. drift is -1.00733 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "lz4"}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#0198137d-1db8-4944-971f-f2d092fa22f8 | FLAKY         | 20/21    | upstream reliability is '97.96672828096118'. current run reliability is '95.23809523809523'. drift is 2.72863 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "lz4"}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a119-4f4e-906a-2327896de76e | FLAKY         | 20/21    | upstream reliability is '97.96672828096118'. current run reliability is '95.23809523809523'. drift is 2.72863 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "snappy"}       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#0198137d-1db8-4c29-ba03-08c5203858db | FLAKY         | 20/21    | upstream reliability is '97.96672828096118'. current run reliability is '95.23809523809523'. drift is 2.72863 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "snappy"}       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a119-4bb2-9337-ad80d219bac8 | FLAKY         | 20/21    | upstream reliability is '97.97047970479706'. current run reliability is '95.23809523809523'. drift is 2.73238 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "zstd"}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#0198137d-1db9-4ca9-8902-8bb43d7f9fc4 | FLAKY         | 20/21    | upstream reliability is '97.96672828096118'. current run reliability is '95.23809523809523'. drift is 2.72863 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest          | test_upgrade_java_compression | {"compression_type": "zstd"}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a11a-453f-a02d-8352572c9c17 | FLAKY         | 20/21    | upstream reliability is '97.96672828096118'. current run reliability is '95.23809523809523'. drift is 2.72863 and the allowed drift is set to 50. The test should PASS  |
| MaintenanceTest              | test_maintenance_sticky       | {"use_rpk": true}                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69092#01981383-a117-41e3-b64d-801ee17e06e4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
</details>
<!--/body-->

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -0,0 +1,128 @@
+ * FetchResponse is buffered per broker. After the response is consumed through
+ * `fetch_next()` it triggers a fetch request.
+ */
+class direct_consumer {
```
what does "direct" mean in the name?

rockwotj replied:
No consumer group. It's the same terminology franz-go uses

rockwotj replied:
Maybe worth explaining this @mmaslankaprv

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -0,0 +1,128 @@
+     * fetched from and is included in the vector its fetch offset will be
+     * updated
+     */
+    ss::future<> assign_partitions(chunked_vector<topic_assignment>);
```
one deviation from the AK API I think is there is no unassign_partitions(partitions) in AK.. assign(list) effectively replaces the existing assignments .. not saying we should change this.. having unassign_partitions is probably more handy for DR given the leadership changes but just wanted to highlight.

mmaslankaprv committed:
kafka/client: introduced helper to determine retriable errors

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
kc/tests: introduced a direct consumer test using cluster mock

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
kc/tests: added direct consumer fixture test

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bharathv commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -0,0 +1,74 @@
+#pragma once
```
license

bharathv commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -0,0 +1,49 @@
+
```
license

bharathv commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -0,0 +1,49 @@
+          _current_bytes += total_bytes;
+          _queue.push_back(
+            entry{.topics = std::move(data), .total_bytes = total_bytes});
+          _updated.signal();
```
neat.... (using the same cv for both pushes and pops).

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -0,0 +1,186 @@
+#include "kafka/client/direct_consumer/direct_consumer.h"
```
license

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -0,0 +1,148 @@
+
+    topic_partition_map<subscription> _subscriptions;
+    chunked_hash_map<model::topic, chunked_vector<model::partition_id>>
+      _removals;
```
nit: comment on removals

mmaslankaprv replied:
i got rid of them completely

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -0,0 +1,186 @@
+void direct_consumer::update_configuration(configuration cfg) { _config = cfg; }
+
+ss::future<> direct_consumer::update_fetchers() {
+    // Stop fetch loops by acquiring a write lock
```
need to hold a gate?

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -0,0 +1,186 @@
+        } else {
+            ++it;
+        }
+    }
```
was wondering if this code needs to reconcile brokers? (eg: metadata response no longer has a broker etc).

mmaslankaprv replied:
i assumed that when the broker is not in metadata it will no longer be reported as a leader for any of the partitions

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -0,0 +1,186 @@
+
+            if (!leader_id) {
+                ssx::spawn_with_gate(_gate, [this] {
+                    return _cluster->request_metadata_update();
```
I think _cluster should expose a queue_metadata_update() or something like that?

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,538 @@
+            // we do not need to hold the lock after we have pushed the data
+            // to the queue, data order is now locked, we can release the lock
+            // and wait for the space in queue
+            lock.return_all();
```
what exactly is the lock guaranteeing?

mmaslankaprv replied:
There are two locks that are being acquired by the fetcher. One is required for correctness the other one i added to make the concurrency model safer. 
First is the shared read-write lock that is owned by the `direct_consumer` this one is being used to synchronize all fetchers and introduce a kind of stop-the-world barrier when fetcher assigned partitions must be updated. This way we guarantee that partition is only assigned to one fetcher and the fetcher is done fetching the partition (there is no inflight request in any of the fetchers) (we can relax this in future by using multiple locks if we need to). The other lock in fetcher is there just to make sure the partition state is not modified while fetching so it is easier to reason about the concurrency.

mmaslankaprv replied:
I ended up removing this locking scheme. Now it is only a single lock protecting `fetcher` fetch state

bharathv reviewed (commented):
high level comments.. most of the code makes sense actually .. we probably need a lot of concurrency testing

rockwotj commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace kafka::client {
+
+class data_queue {
```
looks similar to https://github.com/redpanda-data/redpanda/blob/dev/src/v/transform/transfer_queue.h#L48


Consider adding light documentation

rockwotj commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -0,0 +1,49 @@
+data_queue::push(chunked_vector<fetched_topic_data> data, size_t total_bytes) {
+    return _updated
+      .wait([this, total_bytes] { return can_insert(total_bytes); })
+      .then([total_bytes, data = std::move(data), this]() mutable {
```
no love for coroutines ? :)

rockwotj commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -0,0 +1,148 @@
+    direct_consumer(cluster& cluster, configuration cfg);
+
+    ~direct_consumer();
+    /**
+     * Stops consumer
+     */
+    ss::future<> stop();
```
I'm a bit confused when I see `stop` without a `start`. Do I have to call this? When do I need to call this? Is a consumer automatically started?

mmaslankaprv replied:
i will add `start()` counterpart

rockwotj commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,538 @@
+}
+
+ss::future<fetcher::state_lock_holder> fetcher::lock_state() {
+    /**
+     * VERY IMPORTANT: we need to hold the read lock on the
+     * _parent->_fetchers_lock before we acquire the state lock. Otherwise we
+     * may experience a deadlock as parent is holding the write lock while
+     * modifying the fetcher state and it is acquired before the state lock.
+     */
```
I get a little nervous when seeing stuff like this - we certainly will need good testing to feel confident about this and others can confidently add features over time (topic IDs, fetch sessions, etc)

mmaslankaprv replied:
I think i found a way to get rid of parent lock

rockwotj commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,538 @@
+
+        // this is the place where the loop will be stopped when the parent
+        // consumer updates fetchers state.
+        auto lock = co_await lock_state();
```
This seems bad that if someone configures high `max_wait_time` then all state updates are blocked until that fetch returns (not to mention the additional socket timeouts and delays on top of making a cross AZ network request).

We need to make that clear in the documentation of `direct_consumer` or prevent this kind of issue.

Maybe the current list of partitions to fetch from can be immutable and since I expect updates to be infrequent we just make a copy everytime and swap it out everytime we need to make an update, then we don't need to hold a lock in the parent? I guess that means the parent needs to filter out old changes that are no longer relevant, which seems better than this trickier locking and timeout issues?

mmaslankaprv replied:
The problem i wanted to avoid is to have to deal with the ordering of replies in the direct_consumer and making sure that the per partition order in preserved in returned data. Using this stop-the-world idea prevents this as only one fetcher may fetch the same partition at any given time. Your point makes sense. I will think a bit more how to do it correctly

michael-redpanda commented on src/v/kafka/client/errors.h:
```diff
@@ -0,0 +1,105 @@
+#include "kafka/protocol/errors.h"
+namespace kafka::client {
+
+inline bool is_retriable_error(kafka::error_code ec) {
```
This already exists: https://github.com/redpanda-data/redpanda/blob/71a99186d7161c48763ba82f35e3244ab44d3226/src/v/kafka/protocol/errors.cc#L230

Is this different?

michael-redpanda replied:
If it's different, can you maybe put a comment in as to why the logic in here is different than in `kafka::protocol`?

michael-redpanda commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -0,0 +1,93 @@
+struct fetched_partition_data {
+    model::partition_id partition_id;
+    kafka::leader_epoch leader_epoch;
+    model::offset high_watermark;
```
do we also need to return the LSO?

michael-redpanda replied:
We may need this for metrics information (mirror lag)

michael-redpanda commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Pops data from the queue. If the queue is empty, it will block until
+     * there is data available or the timeout is reached.
+     * If the timeout is reached, it will return empty fetches.
```
I don't think that's accurate - this function will throw a `ss::condition_variable_timed_out` exception if the timeout expires.

michael-redpanda commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -0,0 +1,73 @@
+     * Configuration setters for the queue.
+     */
+    void set_max_bytes(size_t bytes) { _max_bytes = bytes; }
+    void set_max_count(size_t count) { _max_count = count; }
```
should these notify the condition variables if these values are changed?  Say `push` is blocking and nothing is doing a `pop` at the moment

michael-redpanda commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -0,0 +1,73 @@
+ * The queue can store a single entry that exceeds the size limit but the
+ * prerequisite for that is that the queue is empty.
+ */
+class data_queue {
```
I think this would benefit from some unit testing

michael-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -0,0 +1,163 @@
+        int32_t min_bytes{1};
+        int32_t max_fetch_size{512_KiB};
+        int32_t partition_max_bytes{128_KiB};
+        offset_reset_policy reset_policy = offset_reset_policy::earliest;
```
We need to support starting fetch from a specific timestamp, not just earliest/latest

bharathv replied:
Think this is equivalent for `auto.offset.reset` in Kafka

>We need to support starting fetch from a specific timestamp

wdym by fetch from a timestamp, its offset based, right?

mmaslankaprv replied:
exactly, this is equivalent to offset reset property from Kafka clients. It is still possible to fetch from an arbitrary offset when `offset` field is provided in the assignment

michael-redpanda replied:
No this is different.  The ListOffsets API allows you to list offsets that were produced at a specific timestamp.  A user can select where to start mirroring data, either earliest, latest, or at starting at a specific timestamp.

mmaslankaprv replied:
Here it is used to express the reset policy for the consumer not to express the timestamp for the list offset request

michael-redpanda replied:
OH ok I misunderstood then.  So how would the user of direct_consumer specify where they want to start if the user desires to start at a timequery?

michael-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,548 @@
+     * modifying the fetcher state and it is acquired before the state lock.
+     */
+    auto shared_units = co_await _parent->_fetchers_lock.hold_read_lock(_as);
+    auto state_units = co_await _state_lock.get_units();
```
why is this not abortable?

mmaslankaprv replied:
doesn't have to be as the lock is being broken when the fetcher is stopped

michael-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,548 @@
+    }
+    if (needs_backoff) {
+        co_await _parent->_cluster->request_metadata_update();
+        co_await ss::sleep_abortable(error_backoff, _as);
```
This could throw if the sleep is aborted

mmaslankaprv replied:
yes, this is expected

michael-redpanda replied:
Do we want to catch it or does the caller need to handle that exception?

oleiman replied:
should just get eaten by `ssx::repeat_until_gate_closed` iiuc

michael-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -0,0 +1,548 @@
+    if (version) {
+        co_return std::min(version->max, kafka::fetch_api::max_valid);
+    }
+    // if the version is not supported, we fallback to the minimum version
+    // which is 1, this is the first version of the Fetch API that use the new
+    // batch format
+    co_return api_version{1};
```
I'm of the opinion that if this fails, that remote cluster is in a terrible state and we shouldn't be attempting fetches from it - maybe the direct consumer can fetch the version and if it gets an error it throws/returns an error on start up

michael-redpanda commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -0,0 +1,73 @@
+    size_t _max_count{10};
+    size_t _max_bytes{10_MiB};
+
+    size_t _current_bytes{0};
+    ss::chunked_fifo<entry> _queue;
```
nothing to do now but I can see a need to expose this as some sort of metric

michael-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -0,0 +1,232 @@
+    } catch (ss::condition_variable_timed_out&) {
+        co_return chunked_vector<fetched_topic_data>{};
+    }
+}
```
not sure where it needs to live but I think we need to be able to track amount of data coming in, data rates, etc. for metrics.

There's also a req't for tracking this for billing purposes

bharathv replied:
These hook into the pieces I'm adding.

michael-redpanda commented on src/v/kafka/client/test/direct_consumer_test.cc:
```diff
@@ -0,0 +1,611 @@
+      test_log.info,
+      "Changing leadership from {} to 2",
+      cluster_mock.get_topics()[test_topic][model::partition_id(0)].leader);
+    cluster_mock.get_topics()[test_topic][model::partition_id(0)].leader
+      = model::node_id(2);
```
suggestion - maybe expose this as a function in cluster_mock?

michael-redpanda commented on src/v/kafka/client/test/direct_consumer_test.cc:
```diff
@@ -0,0 +1,611 @@
+    }
+
+    ss::future<kafka::fetch_response>
+    handle_fetch(model::node_id id, kafka::fetch_request req) {
```
do we plan on honoring`throttling_quota_exceeded`?  if not the broker will do the throttling - may be helpful to validate how our locking handles broker controlled throttling

mmaslankaprv replied:
we will have to, i will add it into todo list

michael-redpanda commented on src/v/kafka/client/test/cluster_mock.h:
```diff
@@ -131,7 +153,12 @@ class cluster_mock {
       _broker_api_versions;
+    chunked_hash_map<
+      model::topic,
+      chunked_hash_map<model::partition_id, partition_metadata>>
```
I think the value here should be topic_metadata that contains partition_metadata and:
* topic authorized operations
* topic id

mmaslankaprv replied:
we can always add that when we need to, so far i didn't needed it therefore i added the simplest implementation possible

rockwotj commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -0,0 +1,93 @@
+    latest,
+};
+
+inline std::ostream& operator<<(std::ostream& os, offset_reset_policy p) {
```
nit: use `fmt::formatter`

mmaslankaprv replied:
i find the formatters very verbose, and not introducing any value. We are going to automate the `operator<<` conversion therefore i use them in the current code

rockwotj commented on src/v/kafka/client/test/direct_consumer_fixture_test.cc:
```diff
@@ -0,0 +1,199 @@
+    }
+};
+
+TEST_F(consumer_fixture, TestBasicConsumption) {
```
I feel like we could have more expansive tests for things like making sure this assignment epoch fencing works, etc.

rockwotj reviewed (commented):
Generally LGTM - would love to see more tests however.
